### PR TITLE
[SparkDpp] Support complete types

### LINKED ISSUE #4525: [Spark Load][Feature]Support complete data types
1 support decimal andl largeint
2 add validate logic for char/varchar/decimal
3 check data load from hive with strict mode
4 support decimal/date/datetime aggregator

### PR DESCRIPTION
**Pull request by wangbo**
## Proposed changes
see #4525 
For[Spark Load]
1 support decimal andl largeint
2 add validate logic for char/varchar/decimal
3 check data load from hive with strict mode
4 support decimal/date/datetime aggregator

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on (Fix #4525), and have described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [] I have added tests that prove my fix is effective or that my feature works
- [] If this change need a document change, I have updated the document
- [x] Any dependent changes have been merged

### CODE REVIEW

File: fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java [hunk 1]
```diff
@@ -186,4 +192,61 @@ public boolean parse(String value) {
...
+
+    public static int PRECISION = 27;
+    public static int SCALE = 9;
+
+    private BigDecimal maxValue;
+    private BigDecimal minValue;
+
+    public DecimalParser(EtlJobConfig.EtlColumn etlColumn) {
+        StringBuilder precisionStr = new StringBuilder();
+        for (int i = 0; i < etlColumn.precision; i++) {
```
Comment (morningman) [inline fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java:207]: If the column definition is `k1 decimal(4,3)`, than the value range of `k1` should be 
`[-9.999, 9.999]`

And your code will give: `[-9999.999, 9999.999]`, which is not right.

File: fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java [hunk 2]
```diff
@@ -358,12 +362,44 @@ private void processRollupTree(RollupTreeNode rootNode,
...
+                break;
+            case "CHAR":
+            case "VARCHAR":
+                // TODO(wb) padding char type
+                if (srcValue != null && srcValue.toString().length() > etlColumn.stringLength) {
+                    LOG.warn(String.format("the length of input is too long than schema. column_name:%s,input_str[%s],schema length:%s,actual length:%s",
+                            etlColumn.columnName, row.toString(), etlColumn.stringLength, srcValue.toString().length()));
+                    return false;
+                }
+                break;
```
Comment (morningman) [inline fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java:390]: missing `DEFAULT`?

File: fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java (hunk 1 above)
Comment (wyb) [reply fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java:207]: Better to refer to the logic in DecimalLiteral class.

checkPrecisionAndScale(int precision, int scale)

File: fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java [hunk 3]
```diff
@@ -186,4 +192,61 @@ public boolean parse(String value) {
...
+        return minValue;
+    }
+}
+
+class LargeIntParser extends ColumnParser {
+
+    @Override
+    public boolean parse(String value) {
+        try {
+            BigInteger bigInteger = new BigInteger(value);
```
Comment (wyb) [inline fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java:244]: Check largeint type range

File: fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java [hunk 4]
```diff
@@ -507,17 +558,6 @@ private void processRollupTree(RollupTreeNode rootNode,
...
-        dataframe.createOrReplaceTempView("src_table");
-        StringBuilder selectSqlBuilder = new StringBuilder();
-        selectSqlBuilder.append("select ");
-        for (String name : dstColumnNames) {
-            selectSqlBuilder.append(name + ",");
-        }
-        selectSqlBuilder.deleteCharAt(selectSqlBuilder.length() - 1);
-        selectSqlBuilder.append(" from src_table");
-        String selectSql = selectSqlBuilder.toString();
-        dataframe = spark.sql(selectSql);
```
Comment (wyb) [inline fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java:520]: Don't delete. Source data schema may be different from load table schema.
These code is used for projection and reorder columns according to columns order in table schema.

File: fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java [hunk 5]
```diff
@@ -358,12 +362,44 @@ private void processRollupTree(RollupTreeNode rootNode,
...
+                    LOG.warn(String.format("decimal value is not valid for defination, column=%s, value=%s,precision=%s,scale=%s",
+                            etlColumn.columnName, srcValue.toString(), srcBigDecimal.precision(), srcBigDecimal.scale()));
+                    abnormalRowAcc.add(1);
+                    return false;
+                }
+                break;
+            case "CHAR":
+            case "VARCHAR":
+                // TODO(wb) padding char type
+                if (srcValue != null && srcValue.toString().length() > etlColumn.stringLength) {
```
Comment (wyb) [inline fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java:385]: byte length

File: fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java (hunk 1 above)
Comment (wangbo) [reply fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java:207]: 👌

File: fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java (hunk 2 above)
Comment (wangbo) [reply fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java:390]: we just validate char/varchar and decimal here,so it't not necessary

File: fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java (hunk 3 above)
Comment (wangbo) [reply fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java:244]: >BigInteger constructors and operations throw {@code ArithmeticException} when
 >the result is out of the supported range of
 > -2 Integer.MAX_VALUE}(exclusive) to
 > +2Integer.MAX_VALUE}(exclusive)
So I think add a ArithmeticException here is enough

File: fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java (hunk 5 above)
Comment (wangbo) [reply fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/SparkDpp.java:385]: 👌

**Comment by wangbo** [comment]
1 add ut for columnparser and sparkdpp
2 some little fix

**Review by wyb — DISMISSED** [review]
+1

File: fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java [hunk 6]
```diff
@@ -186,4 +194,67 @@ public boolean parse(String value) {
...
+
+    public static int PRECISION = 27;
+    public static int SCALE = 9;
+
+    private BigDecimal maxValue;
+    private BigDecimal minValue;
+
+    public DecimalParser(EtlJobConfig.EtlColumn etlColumn) {
+        StringBuilder precisionStr = new StringBuilder();
+        for (int i = 0; i < etlColumn.precision - etlColumn.scale; i++) {
```
Comment (morningman) [inline fe/spark-dpp/src/main/java/org/apache/doris/load/loadv2/dpp/ColumnParser.java:209]: Here you can write a map to enumerate all the combinations to improve efficiency.

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**